File: backend/src/app/modules/fleet/services/fleet_role_service.py
from __future__ import annotations

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.modules.accounts.models.user import User
from app.modules.fleet.models.fleet import FLEET_MEMBER_ACTIVE, FLEET_ROLE_ADMIRAL
from app.modules.fleet.models.fleet_membership import FleetMembership
from app.modules.fleet.schemas.fleet_role import FleetRoleCreate, FleetRoleRead, FleetRoleUpdate
from app.modules.permissions.models.role import FleetRoleDefinition


class FleetRoleValidationError(ValueError):
    pass


class FleetRolePermissionError(PermissionError):
    pass


def can_manage_fleet_roles(db: Session, actor: User, fleet_id: int) -> bool:
    if actor.is_admin:
        return True
    membership = db.scalar(
        select(FleetMembership)
        .join(FleetMembership.fleet_role)
        .where(
            FleetMembership.fleet_id == fleet_id,
            FleetMembership.user_id == actor.id,
            FleetMembership.status == FLEET_MEMBER_ACTIVE,
            FleetRoleDefinition.code == FLEET_ROLE_ADMIRAL,
        )
    )
    return membership is not None


def _require_permission(db: Session, actor: User, fleet_id: int) -> None:
    if not can_manage_fleet_roles(db, actor, fleet_id):
        raise FleetRolePermissionError("Fleet admiral access required to manage fleet roles.")


def _member_counts(db: Session) -> dict[int, int]:
    return {
        int(role_id): int(count)
        for role_id, count in db.execute(
            select(FleetMembership.fleet_role_id, func.count(FleetMembership.id)).group_by(
                FleetMembership.fleet_role_id
            )
        ).all()
    }


def _serialize(row: FleetRoleDefinition, counts: dict[int, int]) -> FleetRoleRead:
    return FleetRoleRead(
        id=row.id,
        code=row.code,
        label=row.label,
        rank=row.rank,
        is_leadership=row.is_leadership,
        can_manage_fleet=row.can_manage_fleet,
        can_manage_members=row.can_manage_members,
        is_system=row.is_system,
        is_active=row.is_active,
        member_count=counts.get(row.id, 0),
    )


def list_fleet_roles(
    db: Session,
    *,
    include_inactive: bool = False,
) -> list[FleetRoleRead]:
    query = select(FleetRoleDefinition)
    if not include_inactive:
        query = query.where(FleetRoleDefinition.is_active.is_(True))
    rows = db.scalars(
        query.order_by(FleetRoleDefinition.rank.desc(), FleetRoleDefinition.label.asc())
    ).all()
    counts = _member_counts(db)
    return [_serialize(row, counts) for row in rows]


def create_fleet_role(
    db: Session,
    fleet_id: int,
    payload: FleetRoleCreate,
    actor: User,
) -> FleetRoleRead:
    _require_permission(db, actor, fleet_id)
    if db.scalar(select(FleetRoleDefinition.id).where(FleetRoleDefinition.code == payload.code)):
        raise FleetRoleValidationError("A fleet role with this code already exists.")
    row = FleetRoleDefinition(
        code=payload.code,
        label=payload.label,
        rank=payload.rank,
        is_leadership=payload.is_leadership,
        can_manage_fleet=payload.can_manage_fleet,
        can_manage_members=payload.can_manage_members,
        is_system=False,
        is_active=True,
    )
    db.add(row)
    db.commit()
    db.refresh(row)
    return _serialize(row, {})


def update_fleet_role(
    db: Session,
    fleet_id: int,
    role_id: int,
    payload: FleetRoleUpdate,
    actor: User,
) -> FleetRoleRead | None:
    _require_permission(db, actor, fleet_id)
    row = db.get(FleetRoleDefinition, role_id)
    if row is None:
        return None
    if row.is_system:
        raise FleetRoleValidationError("System fleet roles cannot be changed.")
    data = payload.model_dump(exclude_unset=True)
    for field, value in data.items():
        setattr(row, field, value)
    if row.can_manage_fleet or row.can_manage_members:
        row.is_leadership = True
    if data.get("is_active") is False:
        used = db.scalar(
            select(func.count(FleetMembership.id)).where(FleetMembership.fleet_role_id == row.id)
        )
        if used:
            raise FleetRoleValidationError(
                "Reassign all members before deactivating this fleet role."
            )
    db.commit()
    db.refresh(row)
    return _serialize(row, _member_counts(db))


def delete_fleet_role(
    db: Session,
    fleet_id: int,
    role_id: int,
    actor: User,
) -> bool:
    _require_permission(db, actor, fleet_id)
    row = db.get(FleetRoleDefinition, role_id)
    if row is None:
        return False
    if row.is_system:
        raise FleetRoleValidationError("System fleet roles cannot be deleted.")
    used = db.scalar(
        select(func.count(FleetMembership.id)).where(FleetMembership.fleet_role_id == row.id)
    )
    if used:
        raise FleetRoleValidationError("Reassign all members before deleting this fleet role.")
    db.delete(row)
    db.commit()
    return True

File: spring-api/src/main/java/eu/royalblackwater/api/fleet/FleetRoleService.java
package eu.royalblackwater.api.fleet;

import eu.royalblackwater.api.audit.AuditService;
import eu.royalblackwater.api.contract.FleetRoleCreate;
import eu.royalblackwater.api.contract.FleetRoleRead;
import eu.royalblackwater.api.contract.FleetRoleUpdate;
import eu.royalblackwater.api.persistence.JdbcQueryService;
import eu.royalblackwater.api.persistence.RowValues;
import eu.royalblackwater.api.persistence.SqlUpdate;
import eu.royalblackwater.api.security.AuthenticatedUser;
import java.time.Clock;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;
import static org.springframework.http.HttpStatus.BAD_REQUEST;
import static org.springframework.http.HttpStatus.NOT_FOUND;

@Service
public class FleetRoleService {
    private static final String SELECT = """
            select r.*, (select count(*) from fleet_memberships m where m.fleet_role_id=r.id) member_count
            from fleet_roles r
            """;
    private final JdbcQueryService jdbc;
    private final FleetAccessPolicy policy;
    private final AuditService audit;
    private final Clock clock;

    public FleetRoleService(JdbcQueryService jdbc, FleetAccessPolicy policy, AuditService audit, Clock clock) {
        this.jdbc = jdbc;
        this.policy = policy;
        this.audit = audit;
        this.clock = clock;
    }

    @Transactional(readOnly = true)
    public List<FleetRoleRead> list(boolean includeInactive) {
        String filter = includeInactive ? "" : " where r.is_active=true";
        return jdbc.query(SELECT + filter + " order by r.rank desc, r.label asc", Map.of()).stream()
                .map(FleetRoleService::read).toList();
    }

    @Transactional
    public FleetRoleRead create(long fleetId, FleetRoleCreate payload, AuthenticatedUser actor) {
        policy.requireRoleManager(actor, fleetId);
        String code = normalizeCode(payload.code());
        if (jdbc.count("select count(*) from fleet_roles where code=:code", Map.of("code", code)) > 0) {
            throw bad("A fleet role with this code already exists.");
        }
        boolean manageFleet = Boolean.TRUE.equals(payload.canManageFleet());
        boolean manageMembers = Boolean.TRUE.equals(payload.canManageMembers());
        boolean leadership = Boolean.TRUE.equals(payload.isLeadership()) || manageFleet || manageMembers;
        LocalDateTime now = now();
        long id = jdbc.insertReturningId("""
                insert into fleet_roles
                    (code, label, rank, is_leadership, can_manage_fleet, can_manage_members,
                     is_system, is_active, created_at, updated_at)
                values (:code, :label, :rank, :leadership, :manageFleet, :manageMembers,
                        false, true, :now, :now) returning id
                """, Map.of(
                        "code", code, "label", payload.label().strip(), "rank", payload.rank(),
                        "leadership", leadership, "manageFleet", manageFleet,
                        "manageMembers", manageMembers, "now", now));
        audit.record(actor, "fleet_role", id, "create", "Fleet role “" + payload.label().strip() + "” created.",
                List.of("code", "label", "rank", "is_leadership", "can_manage_fleet", "can_manage_members"));
        return find(id);
    }

    @Transactional
    public FleetRoleRead update(long fleetId, long roleId, FleetRoleUpdate payload, AuthenticatedUser actor) {
        policy.requireRoleManager(actor, fleetId);
        Map<String, Object> existing = raw(roleId);
        if (RowValues.booleanValue(existing, "is_system")) {
            throw bad("System fleet roles cannot be changed.");
        }
        if (Boolean.FALSE.equals(payload.isActive()) && RowValues.longValue(existing, "member_count") > 0) {
            throw bad("Reassign all members before deactivating this fleet role.");
        }
        SqlUpdate update = new SqlUpdate("fleet_roles", "id", roleId);
        if (payload.label() != null) update.set("label", payload.label().strip());
        if (payload.rank() != null) {
            if (payload.rank() < 1 || payload.rank() > 79) throw bad("Fleet role rank must be between 1 and 79.");
            update.set("rank", payload.rank());
        }
        if (payload.isLeadership() != null) update.set("is_leadership", payload.isLeadership());
        if (payload.canManageFleet() != null) update.set("can_manage_fleet", payload.canManageFleet());
        if (payload.canManageMembers() != null) update.set("can_manage_members", payload.canManageMembers());
        if (payload.isActive() != null) update.set("is_active", payload.isActive());
        boolean manageFleet = payload.canManageFleet() != null
                ? payload.canManageFleet() : RowValues.booleanValue(existing, "can_manage_fleet");
        boolean manageMembers = payload.canManageMembers() != null
                ? payload.canManageMembers() : RowValues.booleanValue(existing, "can_manage_members");
        if (manageFleet || manageMembers) update.set("is_leadership", true);
        if (!update.isEmpty()) {
            update.set("updated_at", now());
            jdbc.update(update.sql(), update.parameters());
            audit.record(actor, "fleet_role", roleId, "update", "Fleet role #" + roleId + " updated.",
                    update.columns());
        }
        return find(roleId);
    }

    @Transactional
    public void delete(long fleetId, long roleId, AuthenticatedUser actor) {
        policy.requireRoleManager(actor, fleetId);
        Map<String, Object> existing = raw(roleId);
        if (RowValues.booleanValue(existing, "is_system")) {
            throw bad("System fleet roles cannot be deleted.");
        }
        if (RowValues.longValue(existing, "member_count") > 0) {
            throw bad("Reassign all members before deleting this fleet role.");
        }
        jdbc.update("delete from fleet_roles where id=:id", Map.of("id", roleId));
        audit.record(actor, "fleet_role", roleId, "delete", "Fleet role #" + roleId + " deleted.", List.of());
    }

    private FleetRoleRead find(long roleId) {
        return read(raw(roleId));
    }

    private Map<String, Object> raw(long roleId) {
        return jdbc.optional(SELECT + " where r.id=:id", Map.of("id", roleId))
                .orElseThrow(() -> new ResponseStatusException(NOT_FOUND, "Fleet role not found."));
    }

    private static FleetRoleRead read(Map<String, Object> row) {
        return new FleetRoleRead(
                RowValues.booleanValue(row, "can_manage_fleet"),
                RowValues.booleanValue(row, "can_manage_members"),
                RowValues.requiredString(row, "code"), RowValues.longValue(row, "id"),
                RowValues.booleanValue(row, "is_active"), RowValues.booleanValue(row, "is_leadership"),
                RowValues.booleanValue(row, "is_system"), RowValues.requiredString(row, "label"),
                RowValues.nullableLong(row, "member_count"), RowValues.longValue(row, "rank"));
    }

    private static String normalizeCode(String raw) {
        String code = raw.strip().toLowerCase(Locale.ROOT).replace('-', '_').replace(' ', '_');
        if (!code.matches("[a-z][a-z0-9_]{1,39}")) {
            throw bad("Role code must use lowercase letters, numbers and underscores.");
        }
        return code;
    }

    private LocalDateTime now() {
        return LocalDateTime.ofInstant(clock.instant(), ZoneOffset.UTC);
    }

    private static ResponseStatusException bad(String message) {
        return new ResponseStatusException(BAD_REQUEST, message);
    }
}
